compiler/ir/backend.jvm/src/org/jetbrains/kotlin/backend/jvm/lower/CollectionStubMethodLowering.kt

```diff
@@ -0,0 +1,241 @@
+        val typeList = createTypeList(sourceClass, targetClass)
+        if (typeList.isNullOrEmpty() || !typeList.all { it is IrType }) {
+            throw IllegalStateException()
+        }
+        return mutableClass.typeParameters.map { it.symbol }.zip(typeList as List<IrType>).toMap()
+    }
+
+    // Recursively find type list of corresponding immutable collection class.
+    private fun createTypeList(sourceClass: IrClass, targetClass: IrClass): List<IrTypeArgument>? {
+        // First try to find if current class is a direct subtype of source class
+        return (targetClass.superTypes.filter { target ->
+            sourceClass.symbol == target.classifierOrNull
+        }.firstOrNull() as? IrSimpleType)?.arguments ?:
+        // If not direct subtype, recursively look into super types
+        targetClass.superTypes
+            .mapNotNull { it.classOrNull?.owner }
+            .mapNotNull { createTypeList(sourceClass, it) }
+            .firstOrNull()
+    }
+
+    // TODO: replace with new Type substitution utility
+    private fun IrType.substitute(substitutionMap: Map<IrTypeParameterSymbol, IrType>): IrType {
+        if (this !is IrSimpleType) return this
+
+        substitutionMap[classifier]?.let { return it }
+
+        val newArguments = arguments.map {
+            if (it is IrTypeProjection) {
+                makeTypeProjection(it.type.substitute(substitutionMap), it.variance)
+            } else {
+                it
+            }
+        }
+
+        val newAnnotations = annotations.map { it.deepCopyWithSymbols() }
+        return IrSimpleTypeImpl(
+            classifier,
+            hasQuestionMark,
+            newArguments,
+            newAnnotations
+        )
+    }
+
+    // Compute stubs that should be generated, compare based on signature
+    private fun computeStubsToGenerate(irClass: IrClass): HashSet<IrSimpleFunction> =
+        findRelevantSuperCollectionClasses(irClass)
+
+    private data class StubsForCollectionClass(
+        val readOnlyClass: IrClassSymbol,
+        val mutableClass: IrClassSymbol,
+        val mutableOnlyMethods: Collection<IrSimpleFunction>
+    )
+
+    private val preComputedStubs: Collection<StubsForCollectionClass> by lazy {
+        with(context.ir.symbols) {
+            listOf(
+                collection to mutableCollection,
+                set to mutableSet,
+                list to mutableList,
+                map to mutableMap,
+                mapEntry to mutableMapEntry,
+                iterable to mutableIterable,
+                iterator to mutableIterator,
+                listIterator to mutableListIterator
+            ).map { (readOnlyClass, mutableClass) ->
+                val mutableMethods = mutableListOf<IrSimpleFunction>()
+                val readOnlyMethodSignatures = readOnlyClass.functions.map { it.owner.toSignature() }.toHashSet()
+                for (mutableMethod in mutableClass.functions) {
```

Comment:
val mutableMethods = mutableClass.functions.filter { it.toSignature() !in readOnlyMethodSignatures }